### PR TITLE
Add GitHub Actions workflow for docs build

### PR DESCRIPTION
**Pull request by charlesbluca**
Adds a GitHub Actions workflow triggered on merges to `branch-0.20` to build and publish the UCX-Py documentation through GitHub Pages. This handles the issue of libnuma dependencies which currently make this process difficult with RTD.

cc @quasiben

### CODE REVIEW

File: .github/workflows/docs.yml
Comment (charlesbluca) [inline .github/workflows/docs.yml:5]: Unfortunately, GitHub Actions does not have a way to dynamically select the default branch of a repo, so we would need to change this with every new release cycle.

File: .github/workflows/docs.yml
Comment (charlesbluca) [inline .github/workflows/docs.yml:34]: The build will fail here until we add a (preferably empty) `gh-pages` branch to this repo.

**Review by charlesbluca — COMMENTED** [review]
Some comments; also I'm pretty sure we should be able to speed up builds by caching the environment and using Mamba ([here's an xarray workflow doing that](https://github.com/pydata/xarray/blob/0021cdab91f7466f4be0fb32dae92bf3f8290e19/.github/workflows/ci.yaml#L54-L73)), but I'm not exactly sure if/how we could cache the libnuma dependency.

**Comment by pentschev** [comment]
I would prefer that we keep RTD for UCX-Py, the reason being that [UCX uses RTD](https://openucx.readthedocs.io/en/master/) and our intent is to upstream UCX-Py to their repo/org in the next few months.

**Comment by charlesbluca** [comment]
I'm okay with that - in that case, we might want to pursue either:

- a form of these docs that doesn't require a UCX installation
- the possibility of packaging UCX without libnuma dependencies (like the version used [here](https://github.com/quasiben/ucx-py/blob/fix-rtd/conda/environments/builddocs_py37.yml))

In either case, I saw a GitHub Pages site as a short term solution to keep the docs up to date and centralized in this repo until building them through RTD is possible.

**Comment by pentschev** [comment]
I'm a bit confused, what exactly is the problem today with RTD? We do have the docs up, so I'm guessing that's something to do with automatic updates, is that right?

**Comment by charlesbluca** [comment]
From what @quasiben told me, the problem is that the [docs env](https://github.com/rapidsai/ucx-py/blob/branch-0.20/conda/environments/builddocs_py37.yml) in this repo contains UCX, which depends on libnuma, which I don't think can be installed via Conda. Since RTD doesn't allow you to pre-configure the machine docs are being built on, this means that trying to build the docs from this repo will always result in failure.

The docs that are currently published are being built from:

- https://github.com/quasiben/ucx-py/tree/fix-rtd

Which instead uses a specific version of [UCX with no libnuma dependencies](https://github.com/quasiben/ucx-py/blob/eb6bc0a4a5af7eeab0c85eaab35db72116e36fa1/conda/environments/builddocs_py37.yml#L20) which is hosted from a channel I assume @quasiben created. The docs are then periodically updated whenever he merges upstream changes into that branch.

File: .github/workflows/docs.yml
Comment (ajschmidt8) [reply .github/workflows/docs.yml:5]: For this, you could adapt the lines of code below from `cudf` for `ucx-py`. That script gets run when we create new development branches and the `sed` commands can update this line for you automatically.

https://github.com/rapidsai/cudf/blob/branch-0.20/ci/release/update-version.sh#L44-L75

File: .github/workflows/docs.yml
Comment (ajschmidt8) [reply .github/workflows/docs.yml:34]: I've just created an empty `gh-pages` branch for you. https://github.com/rapidsai/ucx-py/tree/gh-pages

**Review by ajschmidt8 — APPROVED** [review]
Approved. For reference, I did something similiar for the `node-rapids` repo. You can check out my implementation below. In particular, I thought you might be able to use [these two `checkout` actions](https://github.com/rapidsai/node-rapids/blob/main/.github/workflows/docs.yaml#L19-L27) to simplify your second `git clone`.


https://github.com/rapidsai/node-rapids/blob/main/.github/workflows/docs.yaml

**Comment by pentschev** [comment]
Thanks for the detailed explanation @charlesbluca . I understand the problem we're trying to solve now, but I'm still unclear on whether this will resolve the problem or just push it down the road. As we intend to move everything here into OpenUCX, we're probably going to need a way of updating RTD, so I don't think the solution here is a proper long-term one.

**Comment by charlesbluca** [comment]
Agreed - I'm hoping that we might be able to find a way to package UCX without libnuma dependencies so that we can eventually move back to RTD, but until then this just seemed like a better alternative to the current setup.

**Comment by pentschev** [comment]
With this, will we still be able to use RTD or does this mean a complete change to GH Pages now that we'll eventually need to move back again to RTD?

**Comment by charlesbluca** [comment]
Thanks for the review @ajschmidt8! Yeah, those `node-rapids` checkout/commit steps definitely look cleaner, I'll adjust my workflow to use that approach then hopefully this should be good to merge.

**Comment by charlesbluca** [comment]
@pentschev we should be able to keep both simultaneously - I was considering having the RTD copy of the docs redirect to the GH pages version, but maybe a different domain name would be better.

**Comment by pentschev** [comment]
If we can keep both, then I think it's ok, but otherwise ping-ponging users back and forth is not really a great approach and we should first consider what other options we have.

**Comment by charlesbluca** [comment]
@gpucibot merge

**Comment by charlesbluca** [comment]
@gpucibot merge

**Comment by charlesbluca** [comment]
Looks like I can't merge this - @pentschev @ajschmidt8 would one of you mind merging this?

**Comment by pentschev** [comment]
> Looks like I can't merge this - @pentschev @ajschmidt8 would one of you mind merging this?

Can you first confirm that RTD will still be available with this change?

**Comment by pentschev** [comment]
Actually, RTD should not just be available, but kept up-to-date, is that the case?

**Comment by charlesbluca** [comment]
Sure! Since the RTD project is based on @quasiben's repo:

image

That should be unaffected by this merge and the RTD site should remain up. However, it would not be up-to-date - if we wanted the RTD site to mirror this one, we would need @quasiben's repo to regularly pull upstream changes and patch them with the no libnuma version of UCX so that the builds don't fail; that could probably be done with a cron job.

Another option that I'm not entirely sure would work is having the RTD site silently redirect to the GitHub pages one, so that they would always be identical. I know RTD supports redirects, though I'm not entirely sure if this is how they work.

**Comment by pentschev** [comment]
Yes, but then we would be duplicating documentation and they will be out-of-sync. That's precisely what I don't want to do. Given we'll soon upstream this repo and will need to update RTD then, I don't really think going to GH Pages is something we should do.

@quasiben can we please hold on to this change and sync first on how we really want/need to handle docs?

**Comment by charlesbluca** [comment]
After a chat with @pentschev and @quasiben, we have a clear roadmap of how to continue publishing the docs through RTD:

- Have a GitHub Actions workflow triggered on merges to the latest branch that checks it out, patches it to use a no libnuma version of UCX, and pushes it to a RTD specific branch (something like `readthedocs`)
- Change RTD builds to be triggered by pushes to `rapidsai/ucx-py@readthedocs` now

In that case, we won't need this workflow anymore, or the empty `gh-pages` branch (though we will need one for RTD later on).

**Comment by pentschev** [comment]
> * Have a GitHub Actions workflow triggered on merges to the latest branch that checks it out, patches it to use a no libnuma version of UCX, and pushes it to a RTD specific branch (something like `readthedocs`)

Just a brief clarification here, I don't think we need to patch anything. The new `readthedocs` branch would be a fork of the current default `branch-0.20` with the necessary [RTD build file](https://github.com/quasiben/ucx-py/blob/fix-rtd/conda/environments/builddocs_py37.yml#L20). Once a PR is merged into `branch-0.20`, GH Actions would merge `branch-0.20` into the `readthedocs` branch as well, so that RTD is always up-to-date.

**Comment by charlesbluca** [comment]
Thanks for the clarification, @pentschev - opened up a PR doing this #721

**Comment by ajschmidt8** [comment]
> After a chat with @pentschev and @quasiben, we have a clear roadmap of how to continue publishing the docs through RTD:
> 
> * Have a GitHub Actions workflow triggered on merges to the latest branch that checks it out, patches it to use a no libnuma version of UCX, and pushes it to a RTD specific branch (something like `readthedocs`)
> * Change RTD builds to be triggered by pushes to `rapidsai/ucx-py@readthedocs` now
> 
> In that case, we won't need this workflow anymore, or the empty `gh-pages` branch (though we will need one for RTD later on).

can this PR be closed then?

**Comment by charlesbluca** [comment]
> can this PR be closed then?

Yes. I think the resolution here is that we actually don't need either workflow, as RTD builds are working fine with the latest version of UCX.

**Pull request closed without merging**
